asQ/allatoncesystem.py

```diff
@@ -0,0 +1,503 @@
+                                        self.aaos.w_recv)
+
+    @PETSc.Log.EventDecorator()
+    def mult(self, mat, X, Y):
+
+        self.aaos.update(X, wall=self.u, wrecv=self.urecv, blocking=True)
+
+        # Set the flag for the circulant option
+        if self.aaos.circ in ["quasi", "picard"]:
+            self.aaos.Circ.assign(1.0)
+        else:
+            self.aaos.Circ.assign(0.0)
+
+        # assembly stage
+        fd.assemble(fd.action(self.Jform, self.u), tensor=self.F)
+        fd.assemble(fd.action(self.Jform_prev, self.urecv),
+                    tensor=self.F_prev)
+        self.F += self.F_prev
+
+        # unset flag if alpha-circulant approximation only in Jacobian
+        if self.aaos.circ not in ["picard"]:
+            self.aaos.Circ.assign(0.0)
+
+        # Apply boundary conditions
+        # assumes aaos.w_all contains the current state we are
+        # interested in
+        # For Jacobian action we should just return the values in X
+        # at boundary nodes
+        for bc in self.aaos.boundary_conditions_all:
+            bc.homogenize()
+            bc.apply(self.F, u=self.u)
+            bc.restore()
+
+        with self.F.dat.vec_ro as v:
+            v.copy(Y)
+
+
+class AllAtOnceSystem(object):
+    def __init__(self,
+                 ensemble, time_partition,
+                 dt, theta,
+                 form_mass, form_function,
+                 w0, bcs=[],
+                 circ="", alpha=1e-3):
+        """
+        The all-at-once system representing multiple timesteps of a time-dependent finite-element problem.
+
+        :arg ensemble: time-parallel ensemble communicator.
+        :arg time_partition: a list of integers for the number of timesteps stored on each ensemble rank.
+        :arg w0: a Function containing the initial data.
+        :arg bcs: a list of DirichletBC boundary conditions on w0.function_space.
+        """
+
+        # check that the ensemble communicator is set up correctly
+        if isinstance(time_partition, int):
+            time_partition = [time_partition]
+        nsteps = len(time_partition)
+        ensemble_size = ensemble.ensemble_comm.size
+        if nsteps != ensemble_size:
+            raise ValueError(f"Number of timesteps {nsteps} must equal size of ensemble communicator {ensemble_size}")
+
+        self.ensemble = ensemble
+        self.time_partition = time_partition
+        self.time_rank = ensemble.ensemble_comm.rank
+        self.nlocal_timesteps = self.time_partition[self.time_rank]
+
+        self.initial_condition = w0
+        self.function_space = w0.function_space()
+        self.boundary_conditions = bcs
+        self.ncomponents = len(self.function_space.split())
+
+        self.dt = dt
+        self.theta = theta
+
+        self.form_mass = form_mass
+        self.form_function = form_function
+
+        self.circ = circ
+        self.alpha = alpha
+
+        if self.circ == "picard":
```

Comment:
This if statement shouldn't be necessary as we should be updating this Constant in the apply method (since it can take different values in the Jacobian and the residual, and we use the same Constant for both).

Reply:
Ok. So we can just set it to zero here?